src/librustc_mir/interpret/operand.rs

```diff
@@ -584,6 +584,8 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 throw_inval!(TooGeneric),
             ty::ConstKind::Unevaluated(def_id, substs) => {
                 let instance = self.resolve(def_id, substs)?;
+                // FIXME: don't use `const_eval_raw` for regular constants, instead use `const_eval`
+                // which immediately validates the result before we continue with it here.
```

Comment:
We *deliberately* did not validate here...

Reply:
For statics that makes sense, but why don't we for constants and promoteds?

Reply:
Hm, that just seemed more consistent to me? Why would statics be different?

Reply:
I always assumed that statics must have a unique address for each static, even for zsts. I'm not sure if this is true, but considering that even heap allocations of zsts get deduplicated into the integer address, idk really

Reply:
> I always assumed that statics must have a unique address for each static, even for zsts. 

"unique"? Two static ZSTs can be at the same address for sure. Because they are zero-sized, one might be at `0x123400` and the other is allocated adjactently -- so, at the same address.

> considering that even heap allocations of zsts get deduplicated into the integer addre

That's what `exchange_malloc` happens to do, but that is not aprt of the Abstract Machine definition.

Also the question was about const validation so I don't know why you are talking about addresses. ;)

Reply:
I lost the context and thought this was a different post.

Statics are different because they can cause cycles, everything else can't cause cycles and can thus be validated immediately when read.

Reply:
Cycles of statics have to involve a ptr indirection, so validating on *reading* should also be cycle-free.

Reply:
I adjusted this to always go through the `const_eval` query.